src/ctap/data_formats.rs

```diff
@@ -1814,6 +1899,56 @@ mod test {
         assert_eq!(cose_key.algorithm, ES256_ALGORITHM);
     }
 
+    #[test]
+    fn test_from_into_cose_signature() {
+        let mut rng = ThreadRng256 {};
+        let sk = crypto::ecdsa::SecKey::gensk(&mut rng);
+        let dummy_signature = sk.sign_rfc6979::<Sha256>(&[]);
+        let mut bytes = [0; ecdsa::Signature::BYTES_LENGTH];
+        dummy_signature.to_bytes(&mut bytes);
+        let cbor_value = cbor_map! {
+            "alg" => ES256_ALGORITHM,
+            "signature" => bytes,
+        };
+        let cose_signature = CoseSignature::try_from(cbor_value).unwrap();
+        let created_signature = crypto::ecdsa::Signature::try_from(cose_signature).unwrap();
+        let mut created_bytes = [0; ecdsa::Signature::BYTES_LENGTH];
+        created_signature.to_bytes(&mut created_bytes);
+        assert_eq!(bytes[..], created_bytes[..]);
+    }
+
+    #[test]
+    fn test_cose_signature_wrong_algorithm() {
+        let mut rng = ThreadRng256 {};
+        let sk = crypto::ecdsa::SecKey::gensk(&mut rng);
+        let dummy_signature = sk.sign_rfc6979::<Sha256>(&[]);
+        let mut bytes = [0; ecdsa::Signature::BYTES_LENGTH];
+        dummy_signature.to_bytes(&mut bytes);
+        let cbor_value = cbor_map! {
+            "alg" => -1, // unused algorithm
+            "signature" => bytes,
+        };
+        let cose_signature = CoseSignature::try_from(cbor_value).unwrap();
+        let created_signature = crypto::ecdsa::Signature::try_from(cose_signature);
+        // Can not compare directly, since ecdsa::Signature does not implement Debug.
+        assert_eq!(
+            created_signature.err(),
+            Some(Ctap2StatusCode::CTAP2_ERR_UNSUPPORTED_ALGORITHM)
+        );
```

Comment:
Nit: `assert_eq!(foo, Err(Some(...)))` would be more idiomatic I think. Also, why is this returning a `Result<..., Option<Ctap2StatusCode>>` instead of simply a `Result<..., Ctap2StatusCode>`?

Reply:
I need the `.err()` because `ecdsa::Signature` implements neither `Debug` nor `PartialEq`, so the whole `Result` can not be compared against. This is also where the `Option` is introduced, as `err()` turns a `Result<T, E>` into `Option<E>`.

Reply:
Got it!